system/security/plugins/testauthSecurity/testauthSecurity.cpp

```diff
@@ -0,0 +1,591 @@
+        if (isEmptyString(node))
+            return access;
+        IPropertyTree* element = accessTree->queryPropTree(node);
+        if (!element)
+            return access;
+        if (element->hasProp("@access"))
+        {
+            int newAccess = element->getPropInt("@access");
+            if (newAccess < SecAccess_Read)
+                return (SecAccessFlags) newAccess;
+            if (first || newAccess < access) //least access wins
+            {
+                access = (SecAccessFlags) newAccess;
+                first = false;
+            }
+        }
+        return getFileScopeAccess(element, nodes, ++pos, access, first);
+    }
+
+    SecAccessFlags queryECLWUScopeAccess(const char* resource)
+    {
+        int* allow = eclwuScopeAccesses.getValue(resource);
+        return allow ? (SecAccessFlags) *allow : defaulteECLWUScopeAccess;
+    }
+};
+
+typedef CUserAccess* CUserAccessPtr;
+
+class CTestAuthSecurityManager : public CBaseSecurityManager
+{
+    static const SecFeatureSet s_implementedFeatureMask = SMF_QuerySecMgrType | SMF_QuerySecMgrTypeName | SMF_CreateAuthMap |
+                                                          SMF_CreateFeatureMap | SMF_LogoutUser | SMF_GetDescription | SMF_Authorize |
+                                                          SMF_AuthorizeEx_Named | SMF_GetAccessFlagsEx |
+                                                          SMF_AuthorizeFileScope_Named | SMF_AuthorizeWorkUnitScope_Named;
+    static const SecFeatureSet s_safeFeatureMask = s_implementedFeatureMask | SMF_CreateSettingMap;
+
+    MapStringTo<CUserAccess*> userAccessMap;
+
+    void readUserAccess(IPropertyTree* secMgrCfg, IPropertyTree* bindConfig)
+    {
+        StringBuffer s;
+        Owned<IPropertyTreeIterator> iter = secMgrCfg->getElements("useraccess");
+        ForEach(*iter)
+        {
+            IPropertyTree& userSettings = iter->query();
+
+            StringBuffer userName, password;
+            if (!readUserNamePassword(userSettings, userName, password))
+                continue;
+
+            CUserAccessPtr* userAccess = userAccessMap.getValue(userName);
+            if (userAccess)
+                continue;
+
+            Owned<CUserAccess> newUserAccess = new CUserAccess(userName, password);
+            readDefaultAccess(userSettings, newUserAccess);
+            readFeatureAccess(bindConfig->queryProp("@serviceType"), userSettings, newUserAccess);
+            readFileScopeAccess(userSettings, newUserAccess);
+            readECLWUScopeAccess(userSettings, newUserAccess);
+            userAccessMap.setValue(userName.str(), newUserAccess.getClear());
+        }
+    }
+
+    bool readUserNamePassword(IPropertyTree& userSettings, StringBuffer& userName, StringBuffer& password)
+    {
+        const char* secretKey = userSettings.queryProp("@secretkey");
+        if (!isEmptyString(secretKey))
+        {
+            const char* vaultKey = userSettings.queryProp("@vaultkey");
```

Comment:
question: Why does this need a separate vault secret?  @afishbeck am I correct the same key can be used for both, when it is passed to getSecret().

Reply:
The name is actually wrong, it should just be vault.  If there are more than one vaults configured (or a mixture of vault and Kubernetes secrets) the vault name would specify which one to look for the secret in.

The vault name is being passed to the right place below, so the attribute just needs to be renamed.

Reply:
It should also be made clear that the vault property restricts the secret to be resolved in that valut.  I think the default behaviour is to search all vaults.  @afishbeck Is this really needed?

Reply:
Removing vault related code